Created Execution API Spec (#149)
ethereum/execution-apis: specs/executionAPISpec.md

|nonce| An integer corresponding to the number of transactions an account has made on the blockchain.|
|peer nodes||
|private key||
|public key||
|rewards||
|signing||
|Smart Contract| Code|
|state||
|storage slot||
|token||
|transaction||
|Uncle block| A block that was |
|user||
|Wei|The smallest unit of Ether where **1Eth = 1e18Wei**|
## 1.3 References
### 1.3.1 Normative References
### 1.3.2 Non-normative References
## 1.4 Document Conventions
### 1.4.1 Input Parameters
Input parameters will be displayed like `this` when referred to in the document.
### 1.4.2 Not Specified vs Null vs Empty
* Not specified describes when a parameter is not part of the call.
* Null describes when a parameter is part of the call, but has the value of null.
* Empty describes when a parameter is part of the call, but has only an empty string.

An example where `gas` is not specified `value` is empty, and `data` is null.
```
{
	"jsonrpc": "2.0",
	"method": "eth_sendTransaction",
	"params": [
		{
			"from": "0x3a5509015e0193adf435a761a6ce160f900034b5",
			"to": "0xe64fac7f3df5ab44333ad3d3eb3fb68be43f2e8c",
			"value": "",
			"data": null
		}
	],
	"id": 1
}
```
### 1.4.3 Default Block Parameter

> Some endpoints use the `defaultBlockParameter` to specify the block that the data is being requested from.
> The `defaultBlockParameter` allows the following options to specify a block:

Reviewer: ```suggestion
Some endpoints use the `blockParameter` to specify the block that the data is being requested from.
The `blockParameter` allows the following options to specify a block:
```
This variable should be renamed throughout the document.  `default` prefix here doesn't make sense, as it is a type and types don't have defaults.

* <a href="#getTransactionByBlockHashAndIndex">eth_getTransactionByBlockHashAndIndex</a>
* <a href="#getTransactionByBlockNumberAndIndex">eth_getTransactionByBlockNumberAndIndex</a>
* <a href="#getRawTransactionByHash">eth_getRawTransactionByHash</a>
* <a href="#getRawTransactionByBlockHashAndIndex">eth_getRawTransactionByBlockHashAndIndex</a>
* <a href="#getRawTransactionByBlockNumberAndIndex">eth_getRawTransactionByBlockNumberAndIndex</a>
* <a href="#getUncleCountByHash">eth_getUncleCountByHash</a> 
* <a href="#getUncleCountByNumber">eth_getUncleCountByNumber</a>
* <a href="#getUncleByBlockHashAndIndex">eth_getUncleByBlockHashAndIndex</a> 
* <a href="#getUncleByBlockNumberAndIndex">eth_getUncleByBlockNumberAndIndex</a>
* <a href="#newFilter">eth_newFilter</a>
* <a href="#newBlockFilter">eth_newBlockFilter</a>
* <a href="#newPendingTransactionFilter">eth_newPendingTransactionFilter</a>
* <a href="#uninstallFilter">eth_uninstallFilter</a>
* <a href="#getFilterChanges">eth_getFilterChanges</a>
* <a href="#getFilterLogs">eth_getFilterLogs</a>
* <a href="#getLogs">eth_getLogs</a>
___
## <p id="clientVersion">web3_clientVersion</p>

> * [WC-1] web3_clientVersion **MUST** return a string containing information about the client version.
> * [WC-2] web3_clientVersion **MUST** start with the name of the client.
> * [WC-3] web3_clientVersion **MUST** return the client version after the name.
> * [WC-4] web3_clientVersion **MUST** return the operating system and architecture the client is running on.
> * [WC-5] web3_clientVersion **MUST** return the name of the language being used.

Reviewer: ```suggestion
* **MUST** return a string containing information about the client version.
* **MUST** start with the name of the client.
* **MUST** return the client version after the name.
* **MUST** return the operating system and architecture the client is running on.
* **MUST** return the name of the language being used.
```
I think the prefix and repeating the method name is unnecessary and makes reading harder in this context.  We already know we are talking about `web3_clientVersion` because that is the section we are in, and the `[WC-#]` doesn't improve readability I don't think.  Maybe the `[WC-#]` is used by the renderer in some way?  If so, then I recommend avoiding abbreviations and using full words so at least we don't have to worry about name collisions.

Reviewer: Same comment for all other sections.

  * The oldest block used for the request.
* [EFH-2] eth_feeHistory **MUST** use the available range of blocks when the requested `blockCount` range can't be retrieved.
* [EFH-3] eth_feeHistory **MUST** allow block tags to be used for `highestBlock`.
* [EFH-4] eth_feeHistory **MUST** error with code -32000 when `highestBlock` is ahead of the chain.
## <p id="blockNumber">eth_blockNumber</p>
* [EBN-1] eth_blockNumber **MUST** return the block number for the most recent block mined.
* [EBN-2] eth_blockNumber **MUST** return "0x0" when the client is not synced to the network.
## <p id="sign">eth_sign</p>
* [ESN-1] eth_sign **MUST** return the Ethereum specific signature detailed in [EIP-191](https://eips.ethereum.org/EIPS/eip-191) for the given unlocked `address` and `message`.
* [ESN-2] eth_sign **MUST** error with code -32000 when the account corresponding to the `address` is not unlocked.
* [ESN-3] eth_sign **MUST** error with code -32000 when the account corresponding to the `address` is not owned by the client.
## <p id="call">eth_call</p>
* [EC-1] eth_call **MUST** return the result of the given transaction.

> * [EC-2] eth_call **MUST** accept all current transaction types. Legacy transactions and [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) "typed" transactions.

Reviewer: ```suggestion
 * [EC-2] eth_call **MUST** accept all transaction types valid for the block that the call is being executed against. This may include legacy transactions or [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) "typed" transactions.
```

> * [EC-3] eth_call **MUST NOT** mine any transaction on the blockchain.

Reviewer: ```suggestion
 * [EC-3] eth_call **MUST NOT** sign nor propagate the transaction (if it happens to be signed) to the network.
```
I think this is what you are getting at, and is more clear what is and isn't allowed by a given client.

> * [EC-4] eth_call **MUST** use the block requested by the `defaultBlockParameter` when interacting with contracts.

Reviewer: ```suggestion
* [EC-4] eth_call **MUST** use the state at the end of the block requested by the `defaultBlockParameter` when interacting with contracts.
```
I actually don't know if pre-state or post-state is used.  You should investigate and find out what each client does and document that here.  I use post-state (state at end of block) as a placeholder above, but that may not be correct, they may use pre-state.

* [EC-5] eth_call **MUST** error with code -32000 when the requested block does not exist or is not available.

> * [EC-6] eth_call **SHOULD NOT** be allowed to be called from an address where CODEHASH != EMPTYCODEHASH. [EIP-3607](https://eips.ethereum.org/EIPS/eip-3607)

Reviewer: I pretty strongly disagree with this one.  Being able to `eth_call` from an address with code is *incredibly* useful during testing and debugging.  EIP-3607 even suggest that clients allow this and list one example use case where it is helpful.

Author: I misread the EIP, would it read better with **SHOULD** because it would ultimately be up to each client if they want to disable it?

Reviewer: If there are not any clients *currently* that disallow `eth_call` from an address with code, then I would say this should be `**MUST** allow`.  It is important that users be able to do this for various reasons, and the only reason I would use `**SHOULD**` is because some client already implemented disallowing it.

> * [EC-7] eth_call **MUST NOT** error if the transaction exceeds the gas or fee cap.

Reviewer: This should probably be two separate lines.  Also, what does it mean for a transaction to "exceed the fee cap"?

Also, it *is* an error for a transaction with a supplied `gas` to exceed that.  Are clients currently ignoring the `gas` provided by the transaction (that sounds like a bug if so, and not a bug we should calcify in a specification).

* [EC-8] eth_call **MUST** use 0x0000000000000000000000000000000000000000 as default `from` address when `from` is null or not specified. [Nethermind uses 0xf...fe] 
* [EC-9] eth_call **MUST** check `from` account balance has sufficient funds to "pay" for the transaction.
* [EC-10] eth_call **MUST** error with code -32000 account has insufficient funds to execute the transaction call.

> * [EC-11] eth_call **MUST NOT** calculate cost of transactions when the `gas` and a price per unit of gas is not specified.   

Reviewer: What does this mean?  `eth_call` doesn't return gas usage numbers currently.  It seems like an implementation detail whether gas account is tracked since it isn't exposed at all.

* [EC-12] eth_call **MUST** allow `data` or `input` to be used when testing deployment or interacting with contracts.
* [EC-13] eth_call **MUST** use `input` when both `input` and `data` are specified. 

> * [EC-14] eth_call **MUST** error with code -32000 when the `data` or `input` being deployed causes and error in the EVM.

Reviewer: This is incorrect.  If there is an EVM failure, then a JSON-RPC success response should be supplied with the revert reason in the result.  At least, this is how clients currently behave.  It is unfortunate, but I am hesitant to make changes to reality in the initial draft of this specification.

> * [EC-15] eth_call **MUST** error with code -32000 when the `gas` is too low to execute the call.

Reviewer: I would need to test to verify, but I *think* this also results in a JSON-RPC success with a result that is `0x`.  If the behavior documented here is what happens in reality, feel free to ignore this comment.

* [EC-16] eth_call **MUST** error with code -32000 when using `gasPrice` with `maxFeePerGas` or `maxPriorityFeePerGas`. 
* [EC-17] eth_call **MUST** use networks chain id for `chainId`

> * [EC-18] eth_call **MUST** error with code -32000 when using block number or block hash for `defaultBlockParameter` while syncing to the network.

Reviewer: This is incorrect.  While it is common for a syncing node to not be able to serve requests and thus will error, this is not a requirement that should be specified.  One can imagine a client that *can* serve requests while syncing.

Also note: Clients are *always* syncing.  I think what you mean here is "during initial sync", but the above comment still applies that this shouldn't be specified.

> * [EC-19] eth_call **MUST** return 0x0 when using block tags for `defaultBlockParameter` while syncing to the network.

Reviewer: I don't understand what this means.  What do you mean "return 0x0"?  Do you mean genesis block should be used when using `latest` or `pending`?  If so, this is incorrect for the same reason as above.  Different clients will have different `latest` blocks while syncing.  Some clients (like Geth) will not have any block available during the initial part of its initial sync, while other clients (like Erigon) will gain access to newer and newer blocks as the sync progresses.

Either way, this isn't something that should be specified, it should be up to the client to return the latest block they have access to when the user asks for `latest`.

* [ECAL-3] eth_createAccessList **MUST** use the estimated gas when the `gas` is not specified.
* [ECAL-4] eth_createAccessList **MUST** error with code -32000 when `gas` is too low.
* [ECAL-5] eth_createAccessList **MUST** error with code -32000 when the `from` address does not have enough Ether to pay for the transaction.
## <p id="estimateGas">eth_estimateGas</p>
* [EEG-1] eth_estimateGas **MUST** return the estimated amount of gas the given `transaction` will take to execute.
* [EEG-2] eth_estimateGas **MUST** error with code -32000 when the `from` address has insufficient Ether to execute the given `transaction`.
* [EEG-3] eth_estimateGas **MUST NOT** check if the `from` account has sufficient funds when estimating contract deployment.
* [EEG-4] eth_estimateGas **MUST** use 0x0000000000000000000000000000000000000000 for `from` when it is not specified.
* [EEG-5] eth_estimateGas **MUST** error with code -32000 when estimating a contract creation that causes an error within the EVM.
## <p id="signTransaction">eth_signTransaction</p>
* [ESNT-1] eth_signTransaction **MUST** return both the signed encoded and signed unencoded format of the `transaction` given.

> * [ESNT-2] eth_signTransaction **MUST NOT** use given transaction `type` for transaction type, **MUST** determine transaction type from given transaction parameters.

Reviewer: This seems very wrong.  If the user specifies `type: 2` in their transaction, a type 2 transaction **MUST** be signed IMO.

If this isn't happening in current clients, we should file a bug with those clients.  It may be valuable to indicate in this specification that if the parameters provided don't align with the transaction type provided by the caller then an error should be returned (with details about that error).

Author: Right now it determines the transaction type by how you pay for gas. when `gasPrice` is used it defaults to Legacy, and when `maxFeePerGas` and `maxPriorityFeePerGas` are used it defaults to type 2 transaction. I will open up a geth issue.

Reviewer: In Geth, if you supply a `type` parameter does it just ignore it completely?

Author: Here is the Geth [issue](https://github.com/ethereum/go-ethereum/issues/24179). Apparently Geth doesn't have a `type` field at all.

* [ESNT-3] eth_signTransaction **MUST** return encoded legacy transactions in the RLP format.
* [ESNT-4] eth_signTransaction **MUST** return encoded [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) transactions as the transaction `type` concatenated with the transaction type's defined encoding method.
* [ESNT-5] eth_signTransaction **MUST** use 0x0000000000000000000000000000000000000000 as default `from` address when `from` is null or not specified.

> * [ESNT-6] eth_signTransaction **MUST** error with code -32000 when the `from` address is locked.

Reviewer: ```suggestion
```
This doesn't make sense.  Accounts cannot be "locked".

> * [ESNT-6] eth_signTransaction **MUST** error with code -32000 when the `from` address is locked.

Reviewer: What does it mean for an account to be "locked"?

> * [ESNT-6] eth_signTransaction **MUST** error with code -32000 when the `from` address is locked.

Reviewer: Side note: Why does *everything* return error code -32000?  It feels like we should be returning useful error codes.  Is this just an issue of how current clients all use the same error code, so now we are stuck with uselessness?

Author: Right! I was wondering the same thing. Most errors do have unique error messages though.

* [ESNT-7] eth_signTransaction **MUST** error with code -32000 when sending transaction using a `from` address that the client does not have the private key for.
* [ESNT-8] eth_signTransaction **MUST** error with code -32000 when the `gas` is not specified.

> * [ESNT-9] eth_signTransaction **MUST** use 0x0 for `gas` when parameter is null.

Reviewer: This results in a useless transaction.  This should either return an error or fill in the gas with a gas estimate, or with a reasonable default.

Reviewer: For things like this, I recommend changing the line to read:
```suggestion
* [ESNT-9] eth_signTransaction **SHOULD** return a JSON-RPC error when the `gas` parameter is null.
```
If current clients don't do it, we can't reasonably put it as a `**MUST**` but it is a behavior that all clients **SHOULD** do.

Author: The document is written to display what Geth currently does. Though while writing this I falsely assumed that using an empty string was the same as using null. Does this sound better?
```
* [ESNT-8] eth_signTransaction **MUST** error with code -32000 when the `gas` is not specified or null.
* [ESNT-9] eth_signTransaction **MUST** use 0x0 for `gas` when empty.
```
I do agree that it should default to something different though.

Reviewer: I think something like this would be better.  Essentially, anytime any of the clients do the wrong thing (like not returning an error on a missing required parameter), we should change "**MUST** do the right thing" to "**SHOULD** do the right thing" so that existing clients aren't out of spec.

Anytime we do **MUST** it should be because:
A. All current clients follow the rule.
B. It is the right behavior.

```
* **MUST** return a JSON-RPC error when the `gas` is not specified or null.  This **SHOULD** be error code -32602.
* **SHOULD** return a JSON-RPC error when the `gas` is an empty string.  This **SHOULD** be error code -32602.
```

* [ESNT-10] eth_signTransaction **MUST** error with code -32000 when the `gasPrice` or `maxFeePerGas` and `maxPriorityFeePerGas` are not specified.
* [ESNT-11] eth_signTransaction **MUST** error with code -32000 when `gasPrice` is used with `maxFeePerGas` and/or `maxPriorityFeePerGas`.

> * [ESNT-12] eth_signTransaction **MUST** use 0x0 for `gasPrice` or `maxFeePerGas` and `maxPriorityFeePerGas` when the parameter is null. 

Reviewer: This will make the transaction unusable, which seems unlikely to be what the user wants.  Either the transaction signing should fail, or a gas price should be chosen by the signer.  I think it is incorrect to prescribe bad behavior in the specification.

Author: Same as below it needs to say `when the parameter is empty` not null. Using null errors which I can add to [ESNT-10]. I agree with you that the spec shouldn't prescribe bad behavior, but I didn't know if I had the authority to declare what I thought the API **should** do. I will write up a doc that lists all the situations like this to make it easier to see what needs to be discussed or changed on the clients.
